agent/consul/state/event.go

```diff
@@ -0,0 +1,82 @@
+package state
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/hashicorp/consul/agent/consul/stream"
+)
+
+type EventPublisher struct {
+	listeners map[stream.Topic]map[*stream.SubscribeRequest]chan stream.Event
+	staged    []stream.Event
+	lock      sync.RWMutex
+}
+
+func NewEventPublisher() *EventPublisher {
+	return &EventPublisher{
+		listeners: make(map[stream.Topic]map[*stream.SubscribeRequest]chan stream.Event),
+	}
+}
+
+// PreparePublish gets an event ready to publish to any listeners on the relevant
+// topics. This doesn't do the send, which happens when the memdb transaction has
+// been committed.
+func (e *EventPublisher) PreparePublish(events []stream.Event) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if e.staged != nil {
+		return errors.New("event already staged for commit")
```

Comment:
Will this prevent publish from ever working after a transaction abort? We should probably either force an explicit abort (which MemDB doesn't have a nice hook for now but we can do it automatically at the end of every transaction in the raft apply handler or something - I think that's what I did in my other branch), or just implicitly abort if this is called a second time and just assume it's not a bug.

If you really want, we could check the raft index of the existing events and only error if it's the same (trying to publish multiple times from the same raft index) and if not then assume an abort happened.

All that said, since we support multiple events anyway we could just append them provided they are distinct topics from the other ones staged? I don't mind any of the above for now - simplest probably best but I think like it is it's incorrect if a transaction ever aborts.